// read-excel-file/modules/xml/parseXmlStream.saxen.js
import { Parser } from 'saxen';
export default function parseXmlStream(xml, _ref) {
  var createInitialState = _ref.createInitialState,
    onOpenTag = _ref.onOpenTag,
    onCloseTag = _ref.onCloseTag,
    onText = _ref.onText;
  // This function was put inside the `parseXmlStream()` function body
  // in order to prevent it from becoming an "external dependency"
  // when using `worker-f` package to run this code in a worker.
  var TAG_NAME_PREFIX = /.+:/;
  function trimXmlnsPrefix(tagName) {
    return tagName.replace(TAG_NAME_PREFIX, '');
  }
  return new Promise(function (resolve, reject) {
    var errored = false;
    var xmlns = true;
    var state = createInitialState();

    // an error happened.
    var onerror = function onerror(error) {
      if (errored) {
        return;
      }
      errored = true;
      reject(error);
    };

    // got some text. `text` is the string of text.
    var ontext = function ontext(text) {
      if (onText) {
        onText(text, state);
      }
    };

    // opened a tag. `node` has "name" and "attributes"
    var onopentag = function onopentag(element, decodeEntities, selfClosing, getContext) {
      if (onOpenTag) {
        // * `element.originalName` — The tag name as written in the XML string, retaining the original prefix regardless of the list of pre-configured namespace mappings.
        // * `element.name` — The tag name with the namespace prefix resolved against the list of pre-configured namespace mappings. I.e. the namespace prefix will potentially be replaced with one from the pre-configured namespace map.
        onOpenTag(xmlns ? trimXmlnsPrefix(element.originalName) : element.name, element.attrs, state);
      }
    };

    // closed a tag.
    var onclosetag = function onclosetag(element) {
      if (onCloseTag) {
        // * `element.originalName` — The tag name as written in the XML string, retaining the original prefix regardless of the list of pre-configured namespace mappings.
        // * `element.name` — The tag name with the namespace prefix resolved against the list of pre-configured namespace mappings. I.e. the namespace prefix will potentially be replaced with one from the pre-configured namespace map.
        var tagName = xmlns ? trimXmlnsPrefix(element.originalName) : element.name;
        onCloseTag(tagName, state);
      }
    };

    // `proxy: true` option enables "proxy" mode.
    //
    // In "proxy" mode, `onopentag` and `onclosetag` receive slightly different arguments:
    // * element name is replaced with element object
    // * getAttribute() function is not passed
    //
    var parser = new Parser({
      proxy: true
    });

    // Parse XML "namespaces" (`xmlns` stuff).
    if (xmlns) {
      parser.ns();
    }
    parser.on('error', onerror);
    parser.on('text', ontext);
    parser.on('openTag', onopentag);
    parser.on('closeTag', onclosetag);
    parser.parse(xml);
    if (errored) {
      return;
    }
    resolve(state);
  });
}
