// read-excel-file/commonjs/xml/parseXmlStream.sax.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = parseXmlStream;
var _nodeStream = require("node:stream");
var _sax = _interopRequireDefault(require("sax"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
// For some strange reason, using "streaming" mode results in
// quite a dramatic drop in performance.
// For example, reading a `10 MB` `.xlsx` file in "streaming" mode is about `3.1` secs.
// while in non-"streaming" mode it's about `1.4` secs.
// Because of such strange performance issue, "streaming" mode is turned off.
var STREAMING_MODE = false;
function parseXmlStream(xml, _ref) {
  var createInitialState = _ref.createInitialState,
    onOpenTag = _ref.onOpenTag,
    onCloseTag = _ref.onCloseTag,
    onText = _ref.onText;
  return new Promise(function (resolve, reject) {
    var errored = false;
    var strict = true;
    var xmlns = true;
    var parserOptions = {
      lowercase: true,
      xmlns: xmlns
    };
    var prefixedTagNameToUnprefixedTagName = {};
    var state = createInitialState();

    // an error happened.
    var onerror = function onerror(error) {
      if (errored) {
        return;
      }
      errored = true;
      reject(error);
    };

    // got some text. `text` is the string of text.
    var ontext = function ontext(text) {
      if (onText) {
        onText(text, state);
      }
    };

    // opened a tag. `node` has "name" and "attributes"
    var onopentag = function onopentag(node) {
      if (onOpenTag) {
        if (xmlns) {
          prefixedTagNameToUnprefixedTagName[node.name] = node.local;
        }
        onOpenTag(xmlns ? node.local : node.name, xmlns ? getAttributesWithoutXmlnsPrefixes(node.attributes) : node.attributes, state);
      }
    };

    // closed a tag.
    var onclosetag = function onclosetag(name) {
      if (onCloseTag) {
        var tagName = xmlns ? prefixedTagNameToUnprefixedTagName[name] : name;
        if (xmlns && !tagName) {
          throw new Error("Unknown closing tag: ".concat(name));
        }
        onCloseTag(tagName, state);
      }
    };

    // an attribute. `attribute` has "name" and "value"
    var onattribute = function onattribute(attribute) {
      // const name = xmlns ? attribute.local : attribute.name
      // const value = attribute.value
    };

    // parser stream is done, and ready to have more stuff written to it.
    var onend = function onend() {
      if (errored) {
        return;
      }
      resolve(state);
    };
    if (STREAMING_MODE) {
      var parser = _sax["default"].createStream(strict, parserOptions);
      parser.on('error', onerror);
      parser.on('text', ontext);
      parser.on('opentag', onopentag);
      parser.on('closetag', onclosetag);
      parser.on('attribute', onattribute);
      parser.on('end', onend);
      _nodeStream.Readable.from(xml).pipe(parser);
    } else {
      var _parser = _sax["default"].parser(strict, parserOptions);
      _parser.onerror = onerror;
      _parser.ontext = ontext;
      _parser.onopentag = onopentag;
      _parser.onclosetag = onclosetag;
      _parser.onattribute = onattribute;
      _parser.onend = onend;
      _parser.write(xml).end();
    }
  });
}
function getAttributesWithoutXmlnsPrefixes(attributes) {
  return Object.keys(attributes).reduce(function (attributesWithoutPrefixes, nameWithPrefix) {
    attributesWithoutPrefixes[attributes[nameWithPrefix].local] = attributes[nameWithPrefix].value;
    return attributesWithoutPrefixes;
  }, {});
}
