// read-excel-file/commonjs/xlsx/parseSharedStrings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = parseSharedStrings;
var _parseSharedString = require("./parseSharedString.js");
/**
 * Parses `sharedStrings.xml` file.
 * @param {string} content
 * @param {function} parseXmlStream — SAX XML parser.
 * @returns {Promise<string[]>}
 */
function parseSharedStrings(content, parseXmlStream) {
  return parseXmlStream(content, {
    createInitialState: createInitialStateInSharedStrings,
    onOpenTag: onOpenTagInSharedStrings,
    onCloseTag: onCloseTagInSharedStrings,
    onText: onTextInSharedStrings
  }).then(function (state) {
    return state.strings;
  });
}
function createInitialStateInSharedStrings() {
  return {
    si: undefined,
    strings: []
  };
}

// In an XLSX `sharedStrings.xml` file, a <si/> (string item) element can have
// either a single child <t/> element (meaning that "rich formatting" is not used)
// or multiple child <r/> elements (meaning that "rich formatting" is used).
function onOpenTagInSharedStrings(tagName, attributes, state) {
  if (tagName === 'si') {
    state.si = (0, _parseSharedString.createInitialStateInSharedString)();
  } else if (state.si) {
    (0, _parseSharedString.onOpenTagInSharedString)(tagName, attributes, state.si);
  }
}
function onCloseTagInSharedStrings(tagName, state) {
  if (tagName === 'si') {
    state.strings.push(state.si.string);
    state.si = undefined;
  } else if (state.si) {
    (0, _parseSharedString.onCloseTagInSharedString)(tagName, state.si);
  }
}
function onTextInSharedStrings(text, state) {
  if (state.si) {
    (0, _parseSharedString.onTextInSharedString)(text, state.si);
  }
}
