// read-excel-file/commonjs/xlsx/parseSheetData.js
"use strict";

function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createInitialStateInSheetData = createInitialStateInSheetData;
exports["default"] = parseSheetData;
exports.onCloseTagInSheetData = onCloseTagInSheetData;
exports.onOpenTagInSheetData = onOpenTagInSheetData;
exports.onTextInSheetData = onTextInSheetData;
var _parseCell = _interopRequireWildcard(require("./parseCell.js"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function parseSheetData(state, sharedStrings, styles, epoch1904, options) {
  return state.cells.map(function (_ref) {
    var attributes = _ref.attributes,
      value = _ref.value,
      inlineString = _ref.inlineString;
    return (0, _parseCell["default"])(attributes, value, inlineString, sharedStrings, styles, epoch1904, options);
  });
}
function createInitialStateInSheetData() {
  return {
    c: undefined,
    cells: []
  };
}
function onOpenTagInSheetData(tagName, attributes, state) {
  if (tagName === 'c') {
    state.c = (0, _parseCell.createInitialStateInCell)();
    state.c.attributes = attributes;
  } else if (state.c) {
    (0, _parseCell.onOpenTagInCell)(tagName, attributes, state.c);
  }
}
function onCloseTagInSheetData(tagName, state) {
  if (tagName === 'c') {
    state.cells.push({
      attributes: state.c.attributes,
      value: state.c.value,
      inlineString: state.c.inlineString
    });
    state.c = undefined;
  } else if (state.c) {
    (0, _parseCell.onCloseTagInCell)(tagName, state.c);
  }
}
function onTextInSheetData(text, state) {
  if (state.c) {
    (0, _parseCell.onTextInCell)(text, state.c);
  }
}

// Here, it could also parse "merged cells" and then return them in some special way.
// But then it's not clear what should be the way to return such merged cells.
// I.e. should it just return it as a duplicate value in each one of the merged cells?
// Or should it keep the current behavior of only returning the value of the top-most left-most cell
// and then just return `null` for the rest of the cells in a "merged cells" group?
// Perhaps the latter (current) approach is the most sensible one, so there's no need
// to change anything.
//
// const mergedCells = getMergedCellCoordinates(sheetDocument)
// for (const mergedCell of mergedCells) {
//   const [from, to] = mergedCell.split(':').map(parseCellAddress)
//   console.log('Merged Cell.', 'From:', from, 'To:', to)
// }
