// read-excel-file/commonjs/xlsx/parseSpreadsheetContents.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = parseSpreadsheetContentsInWorker;
var _parseSpreadsheetInfo = _interopRequireDefault(require("./parseSpreadsheetInfo.js"));
var _parseFilePaths = _interopRequireDefault(require("./parseFilePaths.js"));
var _parseStyles = _interopRequireDefault(require("./parseStyles.js"));
var _parseSharedStrings = _interopRequireDefault(require("./parseSharedStrings.js"));
var _parseSheet = _interopRequireDefault(require("./parseSheet.js"));
var _parseNumber = _interopRequireDefault(require("./parseNumber.js"));
var _checkpoint = _interopRequireWildcard(require("../utility/checkpoint.js"));
var _isPromise = _interopRequireDefault(require("../utility/isPromise.js"));
var _excluded = ["parseNumber"];
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _createForOfIteratorHelperLoose(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (it) return (it = it.call(o)).next.bind(it); if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; return function () { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }
function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
// For an introduction in reading `.xlsx` files see "The minimum viable XLSX reader":
// https://www.brendanlong.com/the-minimum-viable-xlsx-reader.html
/**
 * Reads data from an `.xlsx` file.
 * @param  {function} parseXmlStream — SAX XML parser.
 * @param  {Record<string,string>} contents - A map of `.xml` files inside the `.xlsx` file (which itself is just a zipped directory).
 * @param  {object} [options]
 * @return {Promise<Sheet[]>}
 */
function parseSpreadsheetContents(parseXmlStream, contents) {
  var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  // Because of how `.xlsx` file contents are defined in the specification,
  // it will have to be read in 3 passes:
  // * First pass — read the actual file paths
  // * Second pass — read "shared strings" and "styles"
  // * Thirs pass — read the sheets data

  (0, _checkpoint["default"])('parse spreadsheet info and file paths');

  // Get spreadsheet info and the paths to files.
  return readFiles(getXmlFilesAtFixedPaths(), contents, parseXmlStream).then(function (_ref) {
    var spreadsheetInfo = _ref.spreadsheetInfo,
      filePaths = _ref.filePaths;
    (0, _checkpoint["default"])('parse "shared strings" and "styles"');

    // Parse "shared strings" and "styles".
    return readFiles(getXmlFilesAtNonFixedPaths(filePaths), contents, parseXmlStream).then(function (_ref2) {
      var sharedStrings = _ref2.sharedStrings,
        styles = _ref2.styles;
      var sheetRelationIdsToRead = options.sheets ? options.sheets.map(function (sheet) {
        return getSheetRelationId(sheet, spreadsheetInfo.sheets);
      }) : spreadsheetInfo.sheets.map(function (_) {
        return _.relationId;
      });
      (0, _checkpoint["default"])("parse sheet".concat(sheetRelationIdsToRead.length === 1 ? '' : 's', " data"));

      // Parse sheets data.
      return readFiles(getSheetDataXmlFiles(filePaths, sheetRelationIdsToRead, {
        sharedStrings: sharedStrings,
        styles: styles,
        epoch1904: spreadsheetInfo.epoch1904,
        options: options
      }), contents, parseXmlStream).then(function (sheetsData) {
        (0, _checkpoint["default"])('end');
        // Return sheets data.
        return sheetRelationIdsToRead.map(function (sheetRelationId) {
          return {
            sheet: getSheetNameByRelationId(sheetRelationId, spreadsheetInfo.sheets),
            data: sheetsData[sheetRelationId]
          };
        });
      });
    });
  });
}

/**
 * Reads data from an `.xlsx` file in a worker.
 * @param  {function} [createWorkerFunction] — Creates a worker function.
 * @param  {function} parseXmlStream — SAX XML parser.
 * @param  {Record<string,string>} contents - A map of `.xml` files inside the `.xlsx` file (which itself is just a zipped directory).
 * @param  {object} [options]
 * @return {Promise<Sheet[]>}
 */
function parseSpreadsheetContentsInWorker(createWorkerFunction, parseXmlStream, contents, options) {
  // Assign the default `parseNumber()` function in the `options`.
  // The reason is that the worker code requires it to be non-`undefined`.
  // Otherwise, it would throw "parseNumber is not defined".
  if (!(options && options.parseNumber)) {
    options = _objectSpread(_objectSpread({}, options), {}, {
      parseNumber: _parseNumber["default"]
    });
  }

  // Using a worker requires specifying all the top-level variables or functions that it uses.
  // Currently, that list looks a little bit too long so for now workers aren't used for parsing sheet data.
  // See the comment in `createWorkerFunction()` call for more details.
  // createWorkerFunction = undefined

  // If the environment doesn't support "workers", parse spreadsheet contents "synchronously".
  // This will "block" the main thread while parsing.
  if (!createWorkerFunction) {
    return parseSpreadsheetContents(parseXmlStream, contents, options);
  }

  // Any functions have to be removed from the `options` in order for them to be "serializable"
  // before sending them to the worker thread.
  var _options = options,
    parseNumber = _options.parseNumber,
    optionsJson = _objectWithoutProperties(_options, _excluded);

  // Create a worker.
  var workerFn = createWorkerFunction(function (data) {
    // Reconstruct the `options`.
    var options = _objectSpread(_objectSpread({}, data.optionsJson), {}, {
      parseNumber: parseNumber
    });
    // Parse sheet data from the `.xml` files.
    return parseSpreadsheetContents(parseXmlStream, data.contents, options);
  });
  workerFn.addDependencies(
  // Any "outside" dependencies that're referenced in the function (below).
  function () {
    return [parseXmlStream, parseNumber, parseSpreadsheetContents, getSheetRelationId, getSheetNameByRelationId, getXmlFilesAtFixedPaths, getXmlFilesAtNonFixedPaths, getSheetDataXmlFiles, readFiles, _checkpoint["default"], _checkpoint.latestCheckpointTimestamp, _isPromise["default"]
    // parseSheet,
    //
    // This is not the full list by any means. There's more. Quite a lot more of them.
    // It started looking a bit too much so I stopped adding the dependencies here.
    // Now I understand why `fflate`'s source code is all written in a single
    // multi-thousand-line `index.ts` file. The thing is, once one starts splitting
    // the code into modules, they'd have to manually export any top-level variables
    // or functions from those modules and import them here in order to specify them
    // in the list of dependencies. And even if doing so for every top-level variable
    // or function in every imported module doesn't seem like an impossible task,
    // imagine someone refactoring the code later and extracting new top-level
    // variables or functions. Without 100% code coverage requirement, it won't be caught
    // at build time and can only be caught in production, which isn't ideal to say the least.
    // So it seems like users of this package will have to just deal with the "blocking"
    // nature of the sheet data parser, because I won't follow into `fflate`'s steps
    // and rewrite this package as a single `index.js` file.
    // Users of this package will have to manually put their code in a worker
    // in case they'd prefer it to run in a separate thread to prevent "blocking"
    // the main thread when parsing sheet data.
    ];
  });

  // Post a message with some data to the worker
  // so that it starts processing the data
  // and later posts a message back to the main thread
  // with the result of the calculation.
  return workerFn.callOnce({
    optionsJson: optionsJson,
    contents: contents
  }); // (optional) add `transferList` argument.
}

function getSheetRelationId(sheet, sheets) {
  if (typeof sheet === 'string') {
    for (var _iterator = _createForOfIteratorHelperLoose(sheets), _step; !(_step = _iterator()).done;) {
      var _sheet = _step.value;
      if (_sheet.name === sheet) {
        return _sheet.relationId;
      }
    }
    throw new Error("Sheet \"".concat(sheet, "\" not found. Available sheets: ").concat(sheets.map(function (_ref3) {
      var name = _ref3.name;
      return "\"".concat(name, "\"");
    }).join(', ')));
  } else {
    if (sheet <= sheets.length) {
      return sheets[sheet - 1].relationId;
    }
    throw new Error("Sheet number out of bounds: ".concat(sheet, ". Available sheets count: ").concat(sheets.length));
  }
}
function getSheetNameByRelationId(sheetRelationId, sheets) {
  for (var _iterator2 = _createForOfIteratorHelperLoose(sheets), _step2; !(_step2 = _iterator2()).done;) {
    var sheet = _step2.value;
    if (sheet.relationId === sheetRelationId) {
      return sheet.name;
    }
  }
  throw new Error("Sheet relation ID not found: ".concat(sheetRelationId));
}
function getXmlFilesAtFixedPaths() {
  return {
    // Read the paths to certain files inside the `.xlsx` file, which is itself just a `.zip` archive.
    // These paths aren't standardized between different spreadsheet editors.
    // https://github.com/tidyverse/readxl/issues/104
    'xl/_rels/workbook.xml.rels': {
      name: 'filePaths',
      parse: _parseFilePaths["default"]
    },
    // General info on the spreadsheet.
    'xl/workbook.xml': {
      name: 'spreadsheetInfo',
      parse: _parseSpreadsheetInfo["default"]
    }
  };
}
function getXmlFilesAtNonFixedPaths(filePaths) {
  var _ref4;
  return _ref4 = {}, _defineProperty(_ref4, filePaths.sharedStrings || 'xl/sharedStrings.xml', {
    name: 'sharedStrings',
    // `parseSharedStrings()` returns a `Promise`.
    parse: _parseSharedStrings["default"],
    // It seems that "sharedStrings.xml" is not required to exist.
    // For example, that could be the case when a spreadsheet doesn't contain any strings.
    // https://github.com/catamphetamine/read-excel-file/issues/85
    fallback: []
  }), _defineProperty(_ref4, filePaths.styles || 'xl/styles.xml', {
    name: 'styles',
    parse: _parseStyles["default"],
    fallback: {}
  }), _ref4;
}

// Returns the list of sheet data `.xml` files.
function getSheetDataXmlFiles(filePaths, sheetRelationIdsToRead, sheetDataParserParameters) {
  return Object.keys(filePaths.sheets).filter(function (sheetRelationId) {
    return sheetRelationIdsToRead.includes(sheetRelationId);
  }).reduce(function (filesInfo, sheetRelationId) {
    return _objectSpread(_objectSpread({}, filesInfo), {}, _defineProperty({}, filePaths.sheets[sheetRelationId], {
      name: sheetRelationId,
      // `parseSheet()` returns a `Promise`.
      parse: function parse(content, parseXmlStream) {
        return (0, _parseSheet["default"])(content, parseXmlStream, sheetDataParserParameters);
      }
    }));
  }, {});
}

// In case of converting `.zip` file reader from a "read-and-return" one to a "streaming" one,
// this function could be modified to process the files as they come rather than all-at-once.

// Reads files from inside an `.xlsx` archive by file paths.
//
// In case of converting `.zip` file reader from a "read-and-return" one to a "streaming" one,
// this function could be modified to process the files as they come rather than all-at-once.
//
// But there's a catch: inside an `.xlsx` file, some file paths are not fixed
// and are instead defined in "xl/_rels/workbook.xml.rels" file,
// which presents a "chicken and an egg" dilemma: how could one possibly
// read an `.xlsx` file in one go when the order of the files inside it isn't fixed
// and could be random. Most likely, in the majority of cases, "xl/_rels/workbook.xml.rels"
// file is gonna be one of the first in a given `.xlsx` archive, but still it's not guaranteed.
// A solution would be reading an `.xlsx` file in two passes: one pass would be just to read
// the "xl/_rels/workbook.xml.rels" and ignore decompressing anything else,
// and then the second pass would be to read all other files whose paths are now known.
//
// Returns:
// * If none of the `parse()` functions returned a `Promise`, it returns a map of files' contents.
// * If any of the `parse()` functions returned a `Promise`, it returns a `Promise` that resolves to a map of files' contents.
//
function readFiles(filesInfo, contents, parseXmlStream) {
  // Get files' contents.
  var results = {};
  var _loop = function _loop() {
    var filePath = _Object$keys[_i];
    var fileInfo = filesInfo[filePath];
    results[fileInfo.name] = contents[filePath] === undefined ? fileInfo.fallback === undefined ? function () {
      throw new Error("\"".concat(filePath, "\" file not found inside the `.xlsx` file"));
    }() : fileInfo.fallback : fileInfo.parse(contents[filePath], parseXmlStream);
  };
  for (var _i = 0, _Object$keys = Object.keys(filesInfo); _i < _Object$keys.length; _i++) {
    _loop();
  }
  // Resolve any `Promise`s.
  var promises = [];
  var _loop2 = function _loop2() {
    var name = _Object$keys2[_i2];
    if ((0, _isPromise["default"])(results[name])) {
      promises.push(results[name].then(function (result) {
        results[name] = result;
      }));
    }
  };
  for (var _i2 = 0, _Object$keys2 = Object.keys(results); _i2 < _Object$keys2.length; _i2++) {
    _loop2();
  }
  if (promises.length > 0) {
    return Promise.all(promises).then(function () {
      return results;
    });
  }
  return results;
}
