// read-excel-file/modules/xlsx/parseCell.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import parseCellValue from './parseCellValue.js';
import parseCellAddress from './parseCellAddress.js';

// Example of a `<c/>`ell element:
//
// <c>
//    <f>string</f> — formula.
//    <v>string</v> — formula pre-computed value.
//    <is>
//       <t>string</t> — an `inlineStr` string (rather than a "common string" from a dictionary).
//       <r>
//          <rPr>
//            ...
//          </rPr>
//          <t>string</t>
//       </r>
//       <rPh sb="1" eb="1">
//          <t>string</t>
//       </rPh>
//       <phoneticPr fontId="1"/>
//    </is>
//    <extLst>
//       <ext>
//          <!--any element-->
//       </ext>
//    </extLst>
// </c>
//
export default function parseCell(attributes, value, inlineString, sharedStrings, styles, epoch1904, options) {
  var _parseCellAddress = parseCellAddress(attributes.r),
    _parseCellAddress2 = _slicedToArray(_parseCellAddress, 2),
    row = _parseCellAddress2[0],
    column = _parseCellAddress2[1];
  var type = attributes.t;
  var styleId = attributes.s ? Number(attributes.s) : undefined;
  return {
    row: row,
    column: column,
    value: parseCellValue(value, type, {
      inlineString: inlineString,
      styleId: styleId,
      styles: styles,
      sharedStrings: sharedStrings,
      epoch1904: epoch1904,
      options: options
    })
  };
}
export function createInitialStateInCell() {
  return {
    v: false,
    is: false,
    t: false,
    r: false,
    rPh: false,
    value: undefined,
    inlineString: undefined,
    attributes: undefined
  };
}
export function onOpenTagInCell(tagName, attributes, state) {
  if (tagName === 'v') {
    state.v = true;
  } else if (tagName === 'is') {
    // The possible children of <is> are:
    // * <t> (Text): The standard child to hold plain, simple text.
    // * <r> (Rich Text Run): Used for applying different formatting styles (like bold or italic) to specific segments of text within a single cell.
    // * <rPh> (Phonetic Run): Used primarily for East Asian languages to provide phonetic reading/pronunciation data (e.g., furigana in Japanese). It associates a phonetic pronunciation run right alongside the base string text tag <t>.
    // * <phoneticPr> (Phonetic Properties): Defines formatting and settings for the phonetic text.
    state.is = true;
    state.inlineString = '';
  } else if (tagName === 't') {
    state.t = true;
  } else if (tagName === 'r') {
    // The possible children of <r/> are:
    // * <rPr> (Run Properties): The formatting properties for the text (font, size, color, bold, italic, etc.).
    // * <t> (Text): The actual text payload.
    // * <rPh> (Phonetic Run): Phonetic pronunciation guidance (used for East Asian languages like Japanese). It associates a phonetic pronunciation run right alongside the base string text tag <t>.
    state.r = true;
  } else if (tagName === 'rPh') {
    // The possible children of <rPh/> are:
    // * <t> (Text):  Contains the actual phonetic text or reading (usually in Katakana for Japanese) that corresponds to the associated character string.
    state.rPh = true;
  }
}
export function onCloseTagInCell(tagName, state) {
  if (tagName === 'v') {
    state.v = false;
  } else if (tagName === 'is') {
    state.is = false;
  } else if (tagName === 't') {
    state.t = false;
  } else if (tagName === 'r') {
    state.r = false;
  } else if (tagName === 'rPh') {
    state.rPh = false;
  }
}
export function onTextInCell(text, state) {
  if (state.v) {
    state.value = text;
  } else if (state.is) {
    if (state.rPh) {
      // Ignore anything inside `<rPh/>` tags
    } else if (state.t) {
      if (state.r) {
        // An `<r/>` element could contain multiple `<t/>` elements,
        // the text content from all of which should be concatenated.
        state.inlineString += text;
      } else {
        state.inlineString = text;
      }
    }
  }
}
