// read-excel-file/modules/xlsx/parseSheet.js
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
import parseSheetData, { createInitialStateInSheetData, onOpenTagInSheetData, onCloseTagInSheetData, onTextInSheetData } from './parseSheetData.js';
import parseCellAddress from './parseCellAddress.js';
import reconstructSheetDimensionsFromSheetCells from './reconstructSheetDimensionsFromSheetCells.js';
import convertSheetToData2dArray from './convertCellsToData2dArray.js';

/**
 * Parses a `sheet.xml` file.
 * @param {string} content
 * @param {function} parseXmlStream — SAX XML parser.
 * @param {object} options
 * @returns {Promise<SheetData>}
 */
export default function parseSheet(content, parseXmlStream, _ref) {
  var sharedStrings = _ref.sharedStrings,
    styles = _ref.styles,
    epoch1904 = _ref.epoch1904,
    options = _ref.options;
  // When XML parser is finished, it returns a `state`
  // where it has accumulated the cells it encountered during parsing.
  // `getSheetData()` function reads the parser `state` and transforms it to `SheetData`.
  var getSheetData = function getSheetData(state) {
    var cells = parseSheetData(state.sheetData, sharedStrings, styles, epoch1904, options);
    var dimensions = state.dimension ? parseSheetDimensionRef(state.dimension) : reconstructSheetDimensionsFromSheetCells(cells);
    return convertSheetToData2dArray(cells, dimensions);
  };
  return parseXmlStream(content, {
    createInitialState: createInitialStateInSheet,
    onOpenTag: onOpenTagInSheet,
    onCloseTag: onCloseTagInSheet,
    onText: onTextInSheet
  }).then(function (state) {
    return getSheetData(state);
  });
}
function createInitialStateInSheet() {
  return {
    dimension: undefined,
    sheetData: undefined
  };
}
function onOpenTagInSheet(tagName, attributes, state) {
  if (tagName === 'dimension') {
    state.dimension = attributes.ref;
  } else if (tagName === 'sheetData') {
    state.sheetData = createInitialStateInSheetData();
  } else if (state.sheetData) {
    onOpenTagInSheetData(tagName, attributes, state.sheetData);
  }
}
function onCloseTagInSheet(tagName, state) {
  if (state.sheetData) {
    onCloseTagInSheetData(tagName, state.sheetData);
  }
}
function onTextInSheet(text, state) {
  if (state.sheetData) {
    onTextInSheetData(text, state.sheetData);
  }
}

/**
 * Sheet "dimension" defines the spreadsheet area containing all non-empty cells.
 * Any cells outside the "dimension" are considered empty and should be ignored.
 * https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.spreadsheet.sheetdimension?view=openxml-2.8.1
 * @param {string} `ref` — The value of `<dimension ref/>` attribute.
 * @returns {object[]} `undefined` or `[{ row, column }, { row, column }]` — "From row number and column number to row number and column number".
 */
function parseSheetDimensionRef(ref) {
  var dimensions = ref.split(':').map(parseCellAddress).map(function (_ref2) {
    var _ref3 = _slicedToArray(_ref2, 2),
      row = _ref3[0],
      column = _ref3[1];
    return {
      row: row,
      column: column
    };
  });
  // Sometimes there can be just a single cell as a spreadsheet's "dimensions".
  // For example, the default "dimensions" in Apache POI library is "A1",
  // meaning that only the first cell in the spreadsheet is used.
  //
  // A quote from Apache POI library:
  // "Single cell ranges are formatted like single cell references (e.g. 'A1' instead of 'A1:A1')."
  //
  if (dimensions.length === 1) {
    dimensions = [dimensions[0], dimensions[0]];
  }
  return dimensions;
}
